### PR TITLE
Feature/extract method

### PR DESCRIPTION
**Pull request by CWDN**
Hi,

I originally made a package called `php-extract-method`, it allows users to select a piece of code and extract it out into another function (idea taken from PHPStorm). This was alright and had basic functionality but there were a few bugs and sometimes the spacing was off. What I've done is take some of that and move it into this package. But I've made many improvements when doing this and completely rewrote the logic for getting the parameters needed. 

Planning to deprecate my old package if you are happy with this, as this version is much better.
##### Improvements
1. It now tries to guess the variables you want to return from the method and adds them to the end of the new method. (It also handles multiple returns by using `list` function in php
2. Using all your hard work with the php-integrator-base I'm now able to guess the types of the parameters being passed and the return variable. Then use those types on the docblock.
3. Overall improved the consistency when parsing the parameters to set.
4. General performance improvements.

There's still a small feature that I want to add, which is to allow users to set whether they want to pass the parameters in by reference or for it to return the variables like it currently does. But this feature isn't too important so it can be done at a later date.

I'm adding this feature to this package because it seemed to make more sense to me, rather than creating my own package.

Hope this helps and makes sense. If you have any questions please don't hesitate to ask me.

Thanks
Chris

### CODE REVIEW

**Comment by ghost** [comment]
I have to admit I was secretly hoping you would merge your package into this one ;-). I wasn't sure if the base package would provide any benefits, but I'm glad to hear that it did.

I've done a cursory glance at the code and it seems to be alright, (I'll try to do a more in-depth review later.) In the meantime I've taken it for a spin and have the following suggestions (mostly improvements):
- Could you use a transaction for the inserting of the method code and the replacing of the selected text? This way it can be undone in a single undo action (as opposed to two).
- If possible, clicking outside the dialog should probably cancel it. The `GetterSetterProvider` uses the views from `atom-space-pen-views`, which seem to provide this functionality. I see that `SelectionView` extends `View` from `space-pen`. I'm not sure where this functionality is coming from, but perhaps it's possible to extend one of these to get it for free.
- Could you modify the layout of the button bar so that it matches the one from `GetterSetterProvider`? (In order to maintain consistency between the dialogs.)
- When no function name is entered, the dialog should probably highlight the field in an error markup (I think there is something in the styleguide for this) or alternatively cancel and show a notification. This prevents erroneous code from being generated.
- It would be neat if the variable names and descriptions are vertically aligned in the generated `@param` tags.
- When I extract the code `$test = new AnotherClass;`, the docblock mentions `@return [type]`, is this perhaps a bug (on your side or my side)?
- When I extract the code `$a = new \DateTime();`, I get a `@return bool`, even though the return type is `DateTime` in the following sample (perhaps the type is determined from the lines before the extracted code, perhaps it would work to use the end of the selected code to extract instead):

``` php
$a = true;
$a = new DateTime();
```

Additionally, some things open for discussion:
- I wanted to mention the old array syntax, but I noticed a checkbox conveniently pops up to use the new syntax. I'm wondering, however, if it should be reversed (i.e. checked by default or "Use old array syntax") since it's the modern way.
- What should happen when there is no selection? At the moment the current line seems to be used, which is not incorrect. However, when there is no text on the current line you will get an empty function body. Perhaps we could show a(n info) notification when there is nothing to extract. I've found that you can 'abuse' this to create empty methods in a class, which is a neat side-effect, but a similar feature would probably be larger and allow you to do things such as enter the names and types of parameters.

Also, +1 for an additional newline between the `@param` tag, `@return` tag and description in generated code.

I'm open for discussion if there is something you feel differently about. But in any case, thanks for taking the time to add this!

EDIT: Took another not-so-cursory glance at the code and there is nothing that really jumps out at me (apart from one typo), nice work!

File: lib/ExtractMethodProvider.coffee [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        line = activeTextEditor.lineTextForBufferRow(selectedBufferRange.start.row)
+        findSingleTab = new RegExp("(#{tabText})", "g")
+        matches = (line.match(findSingleTab) || []).length
+
+        multipleTabTexts = Array(matches).fill("#{tabText}")
+        findMultipleTab = new RegExp("^" + multipleTabTexts.join(''), "mg")
+
+        # Replacing double indents with one, so it can be shown in the preview
+        # area of panel
+        reducedHighligtedText = highlightedText.replace(findMultipleTab, "#{tabText}")
```
Comment (ghost) [inline lib/ExtractMethodProvider.coffee]: Minor typo here `reducedHighligtedText` (`highlighted`).

**Comment by CWDN** [comment]
### Fixed

> - Could you use a transaction for the inserting of the method code and the replacing of the selected text? This way it can be undone in a single undo action (as opposed to two).
> - If possible, clicking outside the dialog should probably cancel it. The GetterSetterProvider uses the views from atom-space-pen-views, which seem to provide this functionality. I see that SelectionView extends View from space-pen. I'm not sure where this functionality is coming from, but perhaps it's possible to extend one of these to get it for free.

Had to do custom logic to do this. The `SelectionView` just checks if the `TextEditor` loses focus and cancels the modal then.

> - Could you modify the layout of the button bar so that it matches the one from GetterSetterProvider? (In order to maintain consistency between the dialogs.)
> - When no function name is entered, the dialog should probably highlight the field in an error markup (I think there is something in the styleguide for this) or alternatively cancel and show a notification. This prevents erroneous code from being generated.
> - It would be neat if the variable names and descriptions are vertically aligned in the generated `@param` tags.
> - I wanted to mention the old array syntax, but I noticed a checkbox conveniently pops up to use the new syntax. I'm wondering, however, if it should be reversed (i.e. checked by default or "Use old array syntax") since it's the modern way.

Agreed with what you said, so now it is checked by default.

> - What should happen when there is no selection? At the moment the current line seems to be used, which is not incorrect. However, when there is no text on the current line you will get an empty function body. Perhaps we could show a(n info) notification when there is nothing to extract. I've found that you can 'abuse' this to create empty methods in a class, which is a neat side-effect, but a similar feature would probably be larger and allow you to do things such as enter the names and types of parameters.

I'm now throwing a warning to the user saying they need to select some code before extracting.

### Not fixed (Possible problem with `php-integrator-base`)

> - When I extract the code $test = new AnotherClass;, the docblock mentions `@return [type]`, is this perhaps a bug (on your side or my side)?

Not 100% sure of this but, the only reason it would be `[type]` is if `getVariableType` in your Parser returns null.

> - When I extract the code `$a = new \DateTime();`, I get a `@return bool`, even though the return type is DateTime in the following sample (perhaps the type is determined from the lines before the extracted code, perhaps it would work to use the end of the selected code to extract instead):
> 
> ```
> $a = true;
> $a = new DateTime();
> ```

I'm pretty sure this is something to do with the `getVariableType` function because when you get the scopes it seems to hit the first assignment first and then return straight away. 

Sorry I can't be more helpful with these because my setup could be very different to yours. In my testing it seems like the `getVariableType` function is matching against the first variable assignment.

**Comment by ghost** [comment]
Thanks for implementing my suggestions. About the return type, your interpretation of `getVariableType` is correct. Using this example:

``` php
$a = true;
$a = new DateTime(); // Only this line is selected for extraction.

$a = 5;
```

... I think what's happening is (if I understood the code correctly):
1. Because `$a` needs to be extracted and is used after the selected code to extract, the generated method will contain a `return` statement.
2. The parameter to be returned is determined by `ParameterParser.findParameters` using a regular expression scan.
3. `getTypeForParameter` is called to fetch the type of `$a` relative to where it was found.
4. At this point, `getVariableType` from the base package will start looking for a type anywhere before the passed buffer position (which is where `$a` was first found in the selection), so what it's examining is in fact:

``` php
$a = true;
// The rest of the code falls off because it is >= the passed buffer position.
```

The type is now determined by this assignment as it is closest to the passed buffer position.

In contrast, if you were to pass the end of the selection instead, `getVariableType` will be analyzing:

``` php
$a = true;
$a = new DateTime();
```

... and it should determine the type as being `DateTime`. In this case that seems to be correct as the code after the selected code for extraction expects the type to be `DateTime`, but maybe there are cases where this is incorrect?

**Comment by CWDN** [comment]
Ok, you were completely right. So what I've done is the buffer position I'm passing into `getVariableType` is now the parameters range but `+1` to the end row. By doing this it seems to pick up the correct variable type. The only problem I've found with this is if the user reassigns the variable type on the same line. E.g:

```
$catcher = true; $catcher = new DateTime();
```

In this instance `$catcher` will always evaluate to `DateTime` no matter which part of the line you selected. 

Do you think this is much of a worry or something we need to expect?

---

The only problem I've found is that annotations seem to override any type hinting. So for example:

```
/** @var \App\Hooks\Catcher $catcher */
$catcher = $this->catchers[$appName];

$catcher = true; 
$catcher = new DateTime();
```

This will still evaluate to giving me the variable type of `\App\Hook\Catcher`. Now I can understand that the annotation should probably override any assumptions that you make as the programmer is telling you that `$catcher` is `\App\Hooks\Catcher`. But to me it seems silly to think that, as we can 100% say that `$catcher` is no longer `\App\Hooks\Catcher` and is instead a DateTime object. As I'm not sure about the specification of how annotations are meant to work, what do you think?

**Comment by ghost** [comment]
After your latest changes, I took it for a spin for some real-world usage and noticed the following things (open for discussion, of course):
- The docblock contains the correct types for class types, would it be a good idea to also add these types as type hints in the method signature? I also noticed that the types are not relative to the use statements, but that may very well be a bug in the base package or simply a missing feature on my part.
- Not sure about the `[description]` blocks in the docblock as they are not obligatory and they are not automatically selected with multiple cursors like snippets do. If the user does decide to specify them, he will need to manually remove them anyway, so would it perhaps be better to just leave them out entirely?
- Neat idea (not necessary for this pull request, though): allow specifying the method description in the dialog (prevents having to update the placeholder). Leaving the field empty would not add a description at all (nor a placeholder).

A minor bug: trying to extract a catch block with a type hint will put the catch block variable in a parameter of the extracted method:

``` php
try {

} catch (Exception $e) {

}
```

Regarding the part of the line, I don't really think it will be a problem, but it might be something we can solve: what if you used the end of the selection (i.e. the code to extract) instead of the end of the variable match? Would this generate problems in some situations?

About type hinting: you are right, the type there is obviously `DateTime`. Unfortunately there is no (or I don't know of any) specification regarding the annotations or when they should take effect in the local scope. They are in fact inline docblocks on the code level that were adapted for this purpose by IDE's such as PHPStorm (I believe the draft PSR-5 even has a note somewhere saying that it does not apply to these types of annotations). I'll create a ticket in the base package to investigate further how this should behave.

**Comment by CWDN** [comment]
I've fixed the try catch bug, so it should now exclude the exception from the parameters. I've also added the type hints to the parameters. But it's adding the fully qualified namespace as that is what `getVariableType` returns, which I think is correct for it to return the FQNS as I don't think it is `getVariableType` responsibility to manage the use statements. Now I saw in your autocomplete package a function that will add the namespace to the use statements if required. I think this would be invaluable and would allow me to just add the class name to the type hints. It could probably go into the base package, which would then allow other packages to use this feature. 

With the doc block descriptions, I think that adding more text fields to the modal is going to make it too big and probably won't fit on some monitors. If we were to add a check box to say whether to generate the description placeholders. Then if the user does we replicate the functionality in atom/snippets where you can tab jump between the different markers [1], like what the docblockr package does. I think this would be best, if you agree then I will build it.

Looking at the selection problem again, I remembered that I'm extending their selected range to the start and end column on the start and end rows respectively.  The reason I was doing this was so that in case the user didn't select the entire if statement, for example, this would compensate and select the entire line for you. The only kinda of solution I've thought of is having a config option to specify whether to be fuzzy or precise with the selections and then use the end of the selected range when getting variable type. Do you like this solution or is there a better way? 

[1] https://github.com/atom/snippets/blob/master/lib/snippet-expansion.coffee

**Comment by ghost** [comment]
Regarding the types, adding a use statement for them might not be a bad idea. However, I'd like to do some investigation on how to best approach this first (i.e. it may be best to leave this for an additional improvement after this is merged) as automatically adding use statements raises the following questions:
- Do we directly import every type?
- What if an import for A\B already exists, and the type of a parameter is A\B\C, do we still add a use statement for A\B\C, or de we automatically resolve to B\C?

As you indicated, this is something that the base package (even the PHP side) could learn to handle properly in the future.

Adding a checkbox for the descriptions with tab jumping sounds like a good idea. I'm not sure if this will work, but I noticed the standard atom-snippets package [1] provides a service that contains an `insertSnippet` method that might (hopefully) help you to reuse the existing behavior used in `autocomplete-plus` and other packages. All of this functionality will hopefully also pave the road for other features such as #11.

I'm also wondering about the automatic selection correction. On one side I like the convenience it offers, but on the other side I feel like the user should clearly indicate what he or she wants to extract (even if it's wrong, there is always the preview window as validation). If we go the automatic correction route (or support it with a configuration option), it should probably also handle situations such as:

``` php
$var = "
Multiline string.
The selection starts at this line";
$andEnds = 'here';
```

Perhaps it's easiest to just force the user to properly select the code? Do you happen to know how IDE's handle this? (Not that we need to necessarily match this behavior, but it's always good to compare.)

[1] https://github.com/atom/snippets

**Comment by CWDN** [comment]
### New
- Added the snippets support, so now we consume the snippets manager. 
- Added option to not generate the description placeholders in docblocks
- Grouped checkboxes together in view with labels

### Changed
- Removed the assuming selected range and now works off only what the user has selected. 

I found that PHPStorm does some validation before extracting the method on whether it's possible to extract the code and if the code being extracted is valid. Maybe this is something to look at in the future?

**Comment by ghost** [comment]
Thanks for taking the time to improve this. I think I've kept you busy long enough, so I'll merge this now ;-). As you said, we can always improve the remaining details later. Thanks!

**Pull request merged**